core/src/main/scala/com/lightbend/paradox/markdown/Directive.scala

```diff
@@ -371,14 +391,31 @@ case class JavadocDirective(ctx: Writer.Context)
     case (property @ JavadocLinkStyleProperty(pkg), url) => (pkg, variables(property))
   }
 
+  val defaultPackageNameStyle = variables.getOrElse("javadoc.package_name_style", "startWithLowercase")
+  val packagePackageNameStyle: Map[String, String] = variables.collect {
+    case (property @ JavadocPackageNameStyleProperty(pkg), url) => (pkg, variables(property))
+  }
+
   override protected def resolveApiLink(link: String): Url = {
     val levels = link.split("[.]")
     val packages = (1 to levels.init.size).map(levels.take(_).mkString("."))
     val packagesDeepestFirst = packages.reverse
     val baseUrl = packagesDeepestFirst.collectFirst(baseUrls).getOrElse(defaultBaseUrl).resolve()
     val linkStyle = packagesDeepestFirst.collectFirst(packageLinkStyle).getOrElse(rootLinkStyle)
-    url(link, baseUrl, linkStyle)
+    val packageNameStyle = packagesDeepestFirst.collectFirst(packagePackageNameStyle).getOrElse(defaultPackageNameStyle)
+    url(link, baseUrl, linkStyle, packageNameStyle)
+
+  }
 
+  private def dollarDollarToClassDot(s: String) = s.replaceAll("\\$\\$", ".")
+
+  private[markdown] def url(link: String, baseUrl: Url, linkStyle: LinkStyle, packageNameStyle: String): Url = {
+    val url = Url(link).base
+    val path = dollarDollarToClassDot(ApiDocDirective.packageDotsToSlash(url.getPath, packageNameStyle)) + ".html"
```

Comment:
So any names *actually* containing `$$` are now broken - I suppose that should be exotic enough that this is OK.

Reply:
Yes I also think that this might be a very theoretical case. If we want to get rid of it, we can also think of using a non java character, e.g. a `..` (two dots) instead of `$$`. If I', right then this is invalid for any full qualified name.